pytorch_lightning/accelerators/ddp_accelerator.py

```diff
@@ -145,6 +145,7 @@ def train(self):
         results = self.ddp_train(process_idx=self.task_idx, model=model)
         if 'WORLD_SIZE' in os.environ:
             del os.environ['WORLD_SIZE']
+        self.barrier('ddp_end_train')
```

Comment:
IMO we should land the barriers across DDP accelerators first as that'll solve the primary race condition. We're now getting some crashes from this bug so I want to mitigate as quickly as possible. cc @williamFalcon 

We can address the possible rank inconsistency separately and whether/how to broadcast checkpoint state across ranks. Here's how I can see that turning up:
- the checkpoint callback running on rank 0 thinks the best model path is A due to the metrics it sees
- the checkpoint callback running on rank 1 thinks the best model path is B
- the callback is configured with `save_top_k > 1` so not all checkpoints are saved
- The top-K checkpoints fall out of sync across ranks due to differences in local metrics.
- We delete checkpoints outside the top-K from rank=0.
- Later, rank 1 tries to load the checkpoint (based on its local state), but it turns out it's been deleted

I'm conflicted if we should be broadcasting in that case and clearly document it, or if we should have some giant warning that the metrics need to be synced across ranks in distributed training when saving checkpoints using a monitored value

Reply:
should the barrier be done per accelerator? or should it be added to the trainer itself, right after this spot? https://github.com/PyTorchLightning/pytorch-lightning/blob/f6efb712edccfa83d630db3e97493cd36ab71497/pytorch_lightning/trainer/trainer.py#L439-L440

Reply:
Yeah I agree currently the logic is severely broken because of no barrier. From your report it seems like this issue is larger than just DDP (which was my worry), so I think enforcing a barrier in the trainer is important.

The alternative I see is by incorporating it into the teardown, but that makes it less obvious of the importance here. I'll modify the branch to add the barrier call to the trainer, remove the reference to #4223 as I think this will need to be addressed separately and discard the model checkpoint code.